core/src/main/java/com/linecorp/armeria/common/HttpRequest.java

```diff
@@ -355,10 +356,54 @@ default HttpRequest authority(String authority) {
      * the trailing headers of the request is received fully.
      */
     default CompletableFuture<AggregatedHttpMessage> aggregate(EventExecutor executor) {
+        requireNonNull(executor, "executor");
         final CompletableFuture<AggregatedHttpMessage> future = new CompletableFuture<>();
-        final HttpRequestAggregator aggregator = new HttpRequestAggregator(this, future);
+        final HttpRequestAggregator aggregator = new HttpRequestAggregator(this, future, null);
         completionFuture().whenCompleteAsync(aggregator, executor);
         subscribe(aggregator, executor);
         return future;
     }
+
+    /**
+     * Aggregates this request. The returned {@link CompletableFuture} will be notified when the content and
+     * the trailing headers of the request is received fully. {@link AggregatedHttpMessage#content()} will
+     * return a pooled object, and the caller must ensure to release it. If you don't know what this means,
+     * use {@link HttpResponse#aggregate()}.
+     */
+    default CompletableFuture<AggregatedHttpMessage> aggregateWithPooledObjects(ByteBufAllocator alloc) {
+        requireNonNull(alloc, "alloc");
+        final CompletableFuture<AggregatedHttpMessage> future = new CompletableFuture<>();
+        final HttpRequestAggregator aggregator = new HttpRequestAggregator(this, future, alloc);
+        completionFuture().whenComplete(aggregator);
+        subscribe(aggregator, true);
+        return future;
+    }
+
+    /**
+     * Aggregates this request. The returned {@link CompletableFuture} will be notified when the content and
+     * the trailing headers of the request is received fully. {@link AggregatedHttpMessage#content()} will
+     * return a pooled object, and the caller must ensure to release it. If you don't know what this means,
+     * use {@link HttpResponse#aggregate()}.
+     */
+    default CompletableFuture<AggregatedHttpMessage> aggregateWithPooledObjects(RequestContext ctx) {
```

Comment:
How about removing this method? Although it may be convenient, it introduces an unnecessary dependency to the context API.